src/userdb/userdbctl.c

```diff
@@ -1164,6 +1170,310 @@ static int ssh_authorized_keys(int argc, char *argv[], void *userdata) {
+
+                        log_info("Group with name %s and GID " GID_FMT " already exists, not creating group from credential %s", gr->group_name, gr->gid, name);
+                        return 0;
+                }
+                if (r != -ESRCH)
+                        return log_error_errno(r, "Failed to check if group with name %s already exists: %m", gr->group_name);
+
+                r = groupdb_by_gid(gr->gid, /* match= */ NULL, USERDB_SUPPRESS_SHADOW, &m);
+                if (r >= 0) {
+                        if (!streq_ptr(gr->group_name, m->group_name))
+                                return log_error_errno(SYNTHETIC_ERRNO(EEXIST),
+                                                        "Cannot create group %s from credential %s as group name %s is already assigned to GID " GID_FMT,
+                                                        gr->group_name, name, gr->group_name, m->gid);
+
+                        log_info("Group with name %s and GID " GID_FMT " already exists, not creating group from credential %s", gr->group_name, gr->gid, name);
+                        return 0;
+                }
+                if (r != -ESRCH)
+                        return log_error_errno(r, "Failed to check if group with GID " GID_FMT " already exists: %m", gr->gid);
+
+                r = group_record_clone(gr, USER_RECORD_LOAD_MASK_PRIVILEGED|USER_RECORD_LOG, &gr_stripped);
+                if (r < 0)
+                        return r;
+
+                r = group_record_clone(gr, USER_RECORD_EXTRACT_PRIVILEGED|USER_RECORD_EMPTY_OK|USER_RECORD_LOG, &gr_privileged);
+                if (r < 0)
+                        return r;
+
+                fn = strjoin(gr->group_name, ".group");
+                if (!fn)
+                        return log_oom();
+        }
+
+        if (!filename_is_valid(fn))
+                return log_error_errno(SYNTHETIC_ERRNO(EINVAL),
+                                       "Passed credential '%s' would result in invalid filename '%s'.",
+                                       name, fn);
+
+        _cleanup_free_ char *formatted = NULL;
+        r = sd_json_variant_format(ur ? ur_stripped->json : gr_stripped->json, SD_JSON_FORMAT_NEWLINE, &formatted);
+        if (r < 0)
+                return log_error_errno(r, "Failed to format JSON record: %m");
+
+        r = write_string_file_at(userdb_dir_fd, fn, formatted, WRITE_STRING_FILE_CREATE|WRITE_STRING_FILE_ATOMIC);
+        if (r < 0)
+                return log_error_errno(r, "Failed to write JSON record to /etc/userdb/%s: %m", fn);
+
+        _cleanup_free_ char *link = NULL;
+        if (ur) {
+                if (asprintf(&link, UID_FMT ".user", ur->uid) < 0)
+                        return log_oom();
+        } else {
+                assert(gr);
+
+                if (asprintf(&link, GID_FMT ".group", gr->gid) < 0)
+                        return log_oom();
+        }
```

Comment:
Not a matter, but I'd like to move the link name generation earlier around file name generation.